File: app/src/main/java/eu/kanade/tachiyomi/data/source/model/Page.java
package eu.kanade.tachiyomi.data.source.model;

import java.io.Serializable;
import java.util.List;

import eu.kanade.tachiyomi.data.database.models.Chapter;
import eu.kanade.tachiyomi.data.network.ProgressListener;
import rx.subjects.PublishSubject;

public class Page implements ProgressListener {

    private int pageNumber;
    private String url;
    private String imageUrl;
    private transient Chapter chapter;
    private transient String imagePath;
    private transient volatile int status;
    private transient volatile int progress;

    private transient PublishSubject<Integer> statusSubject;

    public static final int QUEUE = 0;
    public static final int LOAD_PAGE = 1;
    public static final int DOWNLOAD_IMAGE = 2;
    public static final int READY = 3;
    public static final int ERROR = 4;

    public Page(int pageNumber, String url, String imageUrl, String imagePath) {
        this.pageNumber = pageNumber;
        this.url = url;
        this.imageUrl = imageUrl;
        this.imagePath = imagePath;
    }

    public Page(int pageNumber, String url) {
        this(pageNumber, url, null, null);
    }

    public int getPageNumber() {
        return pageNumber;
    }

    public String getUrl() {
        return url;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public String getImagePath() {
        return imagePath;
    }

    public void setImagePath(String imagePath) {
        this.imagePath = imagePath;
    }

    public int getStatus() {
        return status;
    }

    public void setStatus(int status) {
        this.status = status;
        if (statusSubject != null)
            statusSubject.onNext(status);
    }

    public int getProgress() {
        return progress;
    }

    public void setProgress(int value) {
        progress = value;
    }

    @Override
    public void update(long bytesRead, long contentLength, boolean done) {
        progress = (int) ((100 * bytesRead) / contentLength);
    }

    public void setStatusSubject(PublishSubject<Integer> subject) {
        this.statusSubject = subject;
    }

    public Chapter getChapter() {
        return chapter;
    }

    public void setChapter(Chapter chapter) {
        this.chapter = chapter;
    }

    public boolean isLastPage() {
        List<Page> chapterPages = chapter.getPages();
        return chapterPages.size() -1 == pageNumber;
    }
}

File: app/src/main/java/eu/kanade/tachiyomi/data/source/model/Page.kt
package eu.kanade.tachiyomi.data.source.model

import eu.kanade.tachiyomi.data.network.ProgressListener
import eu.kanade.tachiyomi.ui.reader.ReaderChapter
import rx.subjects.Subject

class Page(
        val pageNumber: Int,
        val url: String,
        var imageUrl: String? = null,
        @Transient var imagePath: String? = null
) : ProgressListener {

    @Transient lateinit var chapter: ReaderChapter

    @Transient @Volatile var status: Int = 0
        set(value) {
            field = value
            statusSubject?.onNext(value)
        }

    @Transient @Volatile var progress: Int = 0

    @Transient private var statusSubject: Subject<Int, Int>? = null

    override fun update(bytesRead: Long, contentLength: Long, done: Boolean) {
        progress = (100 * bytesRead / contentLength).toInt()
    }

    fun setStatusSubject(subject: Subject<Int, Int>?) {
        this.statusSubject = subject
    }

    companion object {

        const val QUEUE = 0
        const val LOAD_PAGE = 1
        const val DOWNLOAD_IMAGE = 2
        const val READY = 3
        const val ERROR = 4
    }

}
